TST: pyinstaller fails with errors from test_to_yaml in cosmology

First failure log: https://github.com/astropy/astropy/actions/runs/5367534672/jobs/9737772220

Main suspect: https://github.com/astropy/astropy/pull/14979

```
____________________________ TestFLRW.test_to_yaml _____________________________

self = <astropy_tests.cosmology.flrw.tests.test_base.TestFLRW object at 0x7f127ddfcf40>
cosmo = SubFLRW(name="ABCMeta", H0=70.0 km / (Mpc s), Om0=0.27, Ode0=0.73, Tcmb0=3.0 K, Neff=3.04, m_nu=[0. 0. 0.] eV, Ob0=0.03)
to_format = <astropy.cosmology.connect.CosmologyToFormat object at 0x7f1274074550>
xfail_if_not_registered_with_yaml = None

>   ???
E   AssertionError

astropy/cosmology/_io/tests/test_yaml.py:83: AssertionError
__________________________ TestCosmology.test_to_yaml __________________________

self = <astropy_tests.cosmology.tests.test_core.TestCosmology object at 0x7f127cc87ca0>
cosmo = SubCosmology(name="ABCMeta", H0=70.0 km / (Mpc s), Tcmb0=2.7 K, m_nu=0.6 eV)
to_format = <astropy.cosmology.connect.CosmologyToFormat object at 0x7f127d3f9820>
xfail_if_not_registered_with_yaml = None

>   ???
E   AssertionError

astropy/cosmology/_io/tests/test_yaml.py:83: AssertionError
```

--- a/astropy/cosmology/_io/tests/test_yaml.py
+++ b/astropy/cosmology/_io/tests/test_yaml.py
@@ -75,12 +75,14 @@ class ToFromYAMLTestMixin(ToFromTestMixinBase):
 
     # ===============================================================
 
-    def test_to_yaml(self, cosmo, to_format, xfail_if_not_registered_with_yaml):
+    def test_to_yaml(self, cosmo_cls, to_format, xfail_if_not_registered_with_yaml):
         """Test cosmology -> YAML."""
         yml = to_format("yaml")
 
         assert isinstance(yml, str)  # test type
-        assert yml.startswith("!astropy.cosmology.")
+        assert yml.startswith("!" + ".".join(cosmo_cls.__module__.split(".")[:2]))
+        # e.g. "astropy.cosmology" for built-in cosmologies, or "__main__" for the test
+        # SubCosmology class defined in ``astropy.cosmology.tests.test_core``.
 
     def test_from_yaml_default(
         self, cosmo, to_format, from_format, xfail_if_not_registered_with_yaml

--- a/astropy/cosmology/core.py
+++ b/astropy/cosmology/core.py
@@ -131,7 +131,8 @@ class Cosmology(metaclass=abc.ABCMeta):
         # -------------------
         # Registration
 
-        cls._register_cls()
+        if not inspect.isabstract(cls):  # skip abstract classes
+            cls._register_cls()
 
     @classproperty(lazy=True)
     def _init_signature(cls):

--- a/docs/changes/cosmology/14979.feature.rst
+++ b/docs/changes/cosmology/14979.feature.rst
@@ -1,0 +1,2 @@
+All non-abstract Cosmology subclasses are now automatically registered to work with
+Astropy's YAML serialization.

